File: briar-android/src/test/java/org/briarproject/briar/android/contact/InboxThreadMergerTest.java
package org.briarproject.briar.android.contact;

import org.briarproject.briar.api.telegram.TelegramChat;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class InboxThreadMergerTest {

	@Test
	public void testTelegramRowsMapSecondsToMillis() {
		List<InboxThreadItem> items = InboxThreadMerger.merge(
				Collections.emptyList(),
				Collections.singletonList(new TelegramInboxThreadItem(
						new TelegramChat(7L, "chat", 42)
				))
		);

		TelegramInboxThreadItem item =
				(TelegramInboxThreadItem) items.get(0);
		assertEquals(7L, item.getChatId());
		assertEquals("chat", item.getTitle());
		assertEquals(42000L, item.getLatestActivityMillis());
		assertFalse(item.isPreviewLoading());
		assertFalse(item.isLastMessageOutgoing());
		assertEquals("", item.getPreviewText());
		assertEquals(InboxThreadItem.Source.TELEGRAM, item.getSource());
	}

	@Test
	public void testTelegramRowsExposeCleanLatestPreview() {
		TelegramInboxThreadItem item = new TelegramInboxThreadItem(
				new TelegramChat(7L, "chat", 42,
						"synthetic\npreview\ttext")
		);

		assertFalse(item.isPreviewLoading());
		assertFalse(item.isLastMessageOutgoing());
		assertTrue(item.hasPreviewText());
		assertEquals("synthetic preview text", item.getPreviewText());
	}

	@Test
	public void testTelegramRowsExposeOutgoingLatestPreviewDirection() {
		TelegramInboxThreadItem item = new TelegramInboxThreadItem(
				new TelegramChat(7L, "chat", 42,
						"synthetic\npreview\ttext", true)
		);

		assertFalse(item.isPreviewLoading());
		assertTrue(item.isLastMessageOutgoing());
		assertTrue(item.hasPreviewText());
		assertEquals("synthetic preview text", item.getPreviewText());
	}

	@Test
	public void testTelegramRowsExposeEmptyPreviewState() {
		TelegramInboxThreadItem item = new TelegramInboxThreadItem(
				new TelegramChat(7L, "chat", 42)
		);

		assertFalse(item.isPreviewLoading());
		assertFalse(item.isLastMessageOutgoing());
		assertFalse(item.hasPreviewText());
		assertEquals("", item.getPreviewText());
	}

	@Test
	public void testTelegramRowsExposeLoadingState() {
		TelegramInboxThreadItem item =
				new TelegramInboxThreadItem(7L, "chat", 42000L);

		assertTrue(item.isPreviewLoading());
		assertFalse(item.isLastMessageOutgoing());
		assertFalse(item.hasPreviewText());
		assertEquals("", item.getPreviewText());
	}

	@Test
	public void testMixedItemsSortNewestFirst() {
		List<InboxThreadItem> items = new ArrayList<>(Arrays.asList(
				new FakeInboxThreadItem("older", 1L,
						InboxThreadItem.Source.BRIAR),
				new FakeInboxThreadItem("newer", 3L,
						InboxThreadItem.Source.TELEGRAM),
				new FakeInboxThreadItem("middle", 2L,
						InboxThreadItem.Source.BRIAR)
		));

		InboxThreadMerger.sort(items);

		assertEquals("newer", items.get(0).getStableId());
		assertEquals("middle", items.get(1).getStableId());
		assertEquals("older", items.get(2).getStableId());
	}

	private static class FakeInboxThreadItem implements InboxThreadItem {

		private final String stableId;
		private final long latestActivityMillis;
		private final Source source;

		FakeInboxThreadItem(String stableId, long latestActivityMillis,
				Source source) {
			this.stableId = stableId;
			this.latestActivityMillis = latestActivityMillis;
			this.source = source;
		}

		@Override
		public String getStableId() {
			return stableId;
		}

		@Override
		public long getLatestActivityMillis() {
			return latestActivityMillis;
		}

		@Override
		public Source getSource() {
			return source;
		}
	}
}

File: briar-android/src/test/kotlin/org/briarproject/briar/android/contact/InboxThreadMergerTest.kt
package org.briarproject.briar.android.contact

import org.briarproject.briar.api.connector.ConnectorMessageType
import org.briarproject.briar.api.connector.ConnectorSource
import org.briarproject.briar.api.connector.ConnectorSources
import org.briarproject.briar.api.connector.ConnectorThread
import org.junit.Assert.assertEquals
import org.junit.Test

class InboxThreadMergerTest {

	@Test
	fun testTelegramRowsMapSecondsToMillis() {
		val items = InboxThreadMerger.merge(emptyList(), listOf(telegramItem()))

		val item = items[0] as ConnectorInboxThreadItem
		val rowIdentity = Triple(item.connectorThreadId, item.title, item.latestActivityMillis)
		assertEquals(Triple("7", "chat", 42000L), rowIdentity)
		assertPreviewState(item)
		assertEquals(ConnectorSources.TELEGRAM to "telegram:7", item.connectorSource to item.stableId)
	}

	@Test
	fun testTelegramRowsExposeCleanLatestPreview() {
		val item = telegramItem(text = "synthetic\npreview\ttext")

		assertPreviewState(item, previewText = "synthetic preview text")
	}

	@Test
	fun testTelegramRowsExposeOutgoingLatestPreviewDirection() {
		val item = telegramItem(text = "synthetic\npreview\ttext", outgoing = true)

		assertPreviewState(item, previewText = "synthetic preview text", isLastMessageOutgoing = true)
	}

	@Test
	fun testTelegramRowsExposeLoadingState() {
		val item = GenericConnectorInboxThreadItem(
			connectorSource = ConnectorSources.TELEGRAM,
			connectorThreadId = "7",
			title = "chat",
			latestActivityMillis = 42000L,
			previewText = "",
			isLastMessageOutgoing = false,
			isPreviewLoading = true,
			previewType = ConnectorMessageType.TEXT,
		)

		assertPreviewState(item, isPreviewLoading = true)
	}

	@Test
	fun testMergeSortsTelegramAndMatrixRowsBySourceQualifiedStableId() {
		val telegram = GenericConnectorInboxThreadItem(
			ConnectorThread(
				ConnectorSources.TELEGRAM,
				"7",
				"chat",
				42,
				"",
				false,
				ConnectorMessageType.TEXT,
			),
		)
		val matrix = GenericConnectorInboxThreadItem(
			ConnectorThread(
				ConnectorSources.MATRIX,
				"!room:matrix.example.org",
				"room",
				42,
				"",
				false,
				ConnectorMessageType.TEXT,
			),
		)

		val items = InboxThreadMerger.merge(emptyList(), listOf(telegram, matrix))

		assertEquals(
			setOf("telegram:7", "matrix:!room:matrix.example.org"),
			items.map {
				it.stableId
			}.toSet(),
		)
	}

	@Test
	fun testMergeSortsNewestFirst() {
		val items = InboxThreadMerger.merge(
			emptyList(),
			listOf(connectorItem("older", 1L), connectorItem("newer", 3L), connectorItem("middle", 2L)),
		)

		assertEquals(listOf("m:newer", "m:middle", "m:older"), items.map { it.stableId })
	}

	private fun connectorItem(
		id: String,
		latestActivityMillis: Long,
		connectorSource: ConnectorSource = ConnectorSource("m", "Messenger"),
	) = FakeConnectorInboxThreadItem(
		connectorSource = connectorSource,
		connectorThreadId = id,
		latestActivityMillis = latestActivityMillis,
	)

	private fun telegramItem(text: String = "", outgoing: Boolean = false) =
		GenericConnectorInboxThreadItem(
			ConnectorThread(
				ConnectorSources.TELEGRAM,
				"7",
				"chat",
				42,
				text,
				outgoing,
				ConnectorMessageType.TEXT,
			),
		)

	private fun assertPreviewState(
		item: ConnectorInboxThreadItem,
		previewText: String = "",
		isPreviewLoading: Boolean = false,
		isLastMessageOutgoing: Boolean = false,
	) {
		assertEquals(
			listOf(isPreviewLoading, isLastMessageOutgoing, previewText.isNotEmpty(), previewText),
			listOf(
				item.isPreviewLoading,
				item.isLastMessageOutgoing,
				item.hasPreviewText(),
				item.previewText,
			),
		)
	}
}
